src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs

```diff
@@ -112,9 +113,21 @@ private static async Task<int> Collect(CancellationToken ct, IConsole console, i
                         return ErrorCodes.SessionCreationError;
                     }
 
-                    var collectingTask = new Task(() => {
+                    if (duration != -1)
+                    {
+                        durationTimer = new System.Timers.Timer(duration * 1000);
+                        durationTimer.Elapsed += (s, e) => EventPipeClient.StopTracing(processId, sessionId);
```

Comment:
We've already got a shouldExit ManualResetEvent that triggers cancellation. Is there anything that prevents us from reusing it?
```suggestion
                        durationTimer.Elapsed += (s, e) => shouldExit.Set();
```